src/gevent_server.py

```diff
@@ -0,0 +1,153 @@
+    else:
+        return part_list[1]
+
+
+def response_ok(path, uri, file_type):
+    """Create customized HTTP response for route_uri()."""
+    return_path = os.path.join(path, uri)
+    if os.path.isfile(return_path) is False:
+        return four_oh_four()
+    else:
+        if "image" in file_type:
+            open_file = io.open(return_path, 'rb')
+        else:
+            open_file = io.open(return_path)
+        size = os.path.getsize(return_path)
+        read_file = open_file.read()
+        open_file.close()
+        html_response = "HTTP/1.1 200 OK\r\nContent-Type: {}\r\nContent-length: {}\r\n\r\n".format(file_type, size) + read_file
+        return html_response
+
+
+def four_oh_four():
+    """Send 404 not found message."""
+    open_file = io.open("/Users/admin-1/http_server/http-server/webroot/images/404.jpg" , "rb")
+    read_file = open_file.read()
+    open_file.close()
+    html_response = "HTTP/1.1 400 Not Found\r\n Content-Type: image/jpeg\r\n\r\n" + read_file
+    return html_response
+
+
+def generate_html(base_path, uri):
+    """Create html code to display files."""
+    if uri == "/" or uri == "":
+        join_path = base_path
+    else:
+        join_path = os.path.join(base_path, uri)
+    path = os.walk(join_path)
+    walk_path = next(path)
+    show_dir = walk_path[1]
+    show_files = walk_path[2]
+    inject_html_dir = ""
+    inject_html_file = ""
+    # Create html <li> elements of contents of folder
+    if len(show_dir) > 0:
+        for directory in show_dir:
+            inject_html_dir += "<li><a href={}>{}</a></li>".format(os.path.join(uri, directory), directory)
+    else:
+        inject_html_dir = "No Directories"
+    if len(show_files) > 0:
+        for file in show_files:
+            inject_html_file += "<li><a href={}>{}</a></li>".format(os.path.join(uri, file), file)
+    else:
+        inject_html_file = "No Files"
+    return_path = "Directories:<ul>{}</ul></br>Files:<ul>{}</ul>".format(inject_html_dir, inject_html_file)
+    html_response = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n" + return_path
+    print(html_response)
+    return html_response
+
+
+def route_uri(base_path, uri):
+    """Check to see what file type GET is looking for."""
+    # We have to remove the FIRST slash in the requested uri or else it throws off os.path.join()
+    uri_no_slash = uri.replace("/", "", 1)
+    if "." not in uri:
+        return generate_html(base_path, uri_no_slash)
+    elif ".html" in uri:
+        return response_ok(base_path, uri_no_slash, "text/html")
+    elif ".py" in uri:
+        return response_ok(base_path, uri_no_slash, "text/py")
+    elif ".txt" in uri:
+        return response_ok(base_path, uri_no_slash, "text/plain")
+    # images
+    elif ".jpg" in uri:
+        return response_ok(base_path, uri_no_slash, "image/jpeg")
+    elif ".png" in uri:
+        return response_ok(base_path, uri_no_slash, "image/png")
+    else:
+        return four_oh_four()
```

Comment:
All the code in this module up to this point is exactly the same as the code in the `server.py` module and the `gevent/gevent_server.py` module too.  Why are you repeating all these same functions in so many places instead of simply importing them into this module from the place where they are first implemented.  You have three separate implementations of the same functions, which means you have to test all three of them separately or risk that they drift apart and some become broken while others still work.  This is a serious problem with the DRY-ness of your code.